File: C/hardware/hil_read_analog_example/hil_read_analog_example.c
//////////////////////////////////////////////////////////////////
//
// hil_read_analog_example.c - C file
//
// This example reads one sample immediately from four analog input channels.
//
// This example demonstrates the use of the following functions:
//    hil_open
//    hil_read_analog
//    hil_close
//
// Copyright (C) 2008 Quanser Inc.
//////////////////////////////////////////////////////////////////
    
#include "hil_read_analog_example.h"

int main(int argc, char* argv[])
{
    static const char  board_type[]       = "q2_usb";
    static const char  board_identifier[] = "0";
    static char        message[512];

    qsigaction_t action;
    t_card board;
    t_int  result;

    /* Prevent Ctrl+C from stopping the application so hil_close gets called */
    action.sa_handler = SIG_IGN;
    action.sa_flags   = 0;
    qsigemptyset(&action.sa_mask);

    qsigaction(SIGINT, &action, NULL);

    result = hil_open(board_type, board_identifier, &board);
    if (result == 0)
    {
        const t_uint32 channels[] = { 0, 1 };

        #define NUM_CHANNELS    ARRAY_LENGTH(channels)

        t_double voltages[NUM_CHANNELS];

        result = hil_read_analog(board, channels, NUM_CHANNELS, &voltages[0]);
        if (result >= 0)
        {
            t_uint32 channel;
            for (channel = 0; channel < NUM_CHANNELS; channel++)
                printf("ADC #%d: %7.4f   ", channels[channel], voltages[channel]);
            printf("\n");
        }
        else
        {
            msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
            printf("Unable to read channels. %s Error %d.\n", message, -result);
        }
            
        hil_close(board);
    }
    else
    {
        msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
        printf("Unable to open board. %s Error %d.\n", message, -result);
    }

    printf("\nPress Enter to continue.\n");
    getchar();

    return 0;
}

File: python/hardware/hil_read_analog_example/hil_read_analog_example.py
#################################################################
#
# hil_read_analog_example.py - Python file
#
# This example reads one sample immediately from two analog input channels.
#
# This example demonstrates the use of the following functions:
#    HIL.open
#    HIL.read_analog
#    HIL.close
#
# Copyright (C) 2023 Quanser Inc.
#################################################################

from quanser.hardware import HIL, HILError, Clock
import math
import array as arr

board_type = "q2_usb"
board_identifier = "0"

try:
    card = HIL()
    card.open(board_type, board_identifier)

    try:
        channels = arr.array("i", [0, 1])
        num_channels = len(channels)

        voltages = arr.array("d", [0.0] * num_channels)

        card.read_analog(channels, num_channels, voltages)
        for channel in range(num_channels):
            print("ADC #%d: %7.4f" % (channels[channel], voltages[channel]), end="   ")
        print()

    except HILError as ex:
        print("Unable to read channels. %s" % ex.get_error_message())

    finally:
        card.close()

except HILError as ex:
    print("Unable to open board. %s" % ex.get_error_message())

input("Press Enter to continue.")
exit(0)
